core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala

```diff
@@ -0,0 +1,121 @@
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import com.yammer.metrics.Metrics
+import kafka.cluster.BrokerEndPoint
+import kafka.common.TopicAndPartition
+import kafka.message.ByteBufferMessageSet
+import kafka.server.AbstractFetcherThread.{FetchRequest, PartitionData}
+import kafka.utils.TestUtils
+import org.apache.kafka.common.protocol.Errors
+import org.junit.Assert.{assertFalse, assertTrue}
+import org.junit.Test
+
+import scala.collection.JavaConverters._
+
+class AbstractFetcherThreadTest {
+
+  @Test
+  def testMetricsRemovedOnShutdown() {
+    val partition = new TopicAndPartition("topic", 0)
+    val fetcherThread = new DummyFetcherThread("dummy", "client", new BrokerEndPoint(0, "localhost", 9092))
+
+    fetcherThread.start()
+
+    // add one partition to create the consumer lag metric
+    fetcherThread.addPartitions(Map(partition -> 0L))
+
+    // wait until all fetcher metrics are present
+    TestUtils.waitUntilTrue(() =>
+      allMetricsByName == Set(FetcherMetrics.BytesPerSec, FetcherMetrics.RequestsPerSec, FetcherMetrics.ConsumerLag),
+      "Failed waiting for all fetcher metrics to be registered")
+
+    fetcherThread.shutdown()
+
+    // after shutdown, they should be gone
+    assertTrue(Metrics.defaultRegistry().allMetrics().isEmpty)
+  }
+
+  @Test
+  def testConsumerLagRemovedWithPartition() {
+    val partition = new TopicAndPartition("topic", 0)
+    val fetcherThread = new DummyFetcherThread("dummy", "client", new BrokerEndPoint(0, "localhost", 9092))
+
+    fetcherThread.start()
+
+    // add one partition to create the consumer lag metric
+    fetcherThread.addPartitions(Map(partition -> 0L))
+
+    // wait until lag metric is present
+    TestUtils.waitUntilTrue(() => allMetricsByName(FetcherMetrics.ConsumerLag),
+      "Failed waiting for consumer lag metric")
+
+    // remove the partition to simulate leader migration
+    fetcherThread.removePartitions(Set(partition))
+
+    // the lag metric should now be gone
+    assertFalse(allMetricsByName(FetcherMetrics.ConsumerLag))
+
+    fetcherThread.shutdown()
+  }
+
+  private def allMetricsByName = Metrics.defaultRegistry().allMetrics().asScala.keySet.map(_.getName)
```

Comment:
Is this simply `allMetricNames`? The current name makes it sound (to me) like it's a map.